### PR TITLE
Silver solver 6

### PR DESCRIPTION
**Pull request by SilverSolver**
I used crossopt's solution as a baseline and added the validation of input + simplest tests

### CODE REVIEW

File: group/assignment6/move_cars.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+def generate_moves(start, final):
+    """ ith element of start start is the car currently in the ith space
+    ith element of final start is the car that should be in ith space
+    empty space is 0
+    yields optimal sequence of moves in format (car number, old space, new space) """
+
+    # validation of the input
```
Comment (crossopt) [inline group/assignment6/move_cars.py:7]: This should probably be a separate function.
I think I remember something about each part of the code needing to have a separate responsibility?

It also seems logical to have validation be optional. And the validation for opt3 is going to be, mmm, less nontrivial, so makes sense to keep it separate from the beginning and have the opt3 validation reuse this piece

File: group/assignment6/move_cars.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    start_set = set(start)
+    final_set = set(final)
+    #print(start_set, final_set)
+    if len(start_set) != len(start) or start_set != final_set or \
+       len(final_set) != len(final) or \
+       start_set != final_set and 0 not in start_set:
+        yield None
+
+    # pos will contain car positions in another form
+    pos = [0] * len(start)
```
Comment (crossopt) [inline group/assignment6/move_cars.py:17]: Same thing here about code structure. Should this be a logical part of the generation?

My personal preference is to have it in the same function, but, well
Competitive programming: if you can read it 5 minutes later, it's good code

I'm probably not the best judge here:)

File: group/assignment6/move_cars.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                pos[0], pos[car] = pos[car], pos[0]
+                car = final[pos[0]]
+
+def fancy_output(start, final):
+    result = generate_moves(start, final)
+    for i in result:
+        if i:
+            print("Move car number {} from {} to {}".format(*i))
+        else:
+            print("Unsolvable")
```
Comment (crossopt) [inline group/assignment6/move_cars.py:43]: Unsolvable?
More like "Wrong input format"

Opt3 will actually have unsolvable cases (and I'll try to get around to writing it in the next couple of days), so best to keep the distinction

Maybe we want exceptions for wrong input format? Possibly checking the Google code style beforehand

File: group/assignment6/move_cars.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            break
+    print("Done")
+
+def return_output(start, final):
```
Comment (crossopt) [inline group/assignment6/move_cars.py:47]: Why does this function exist?

I added an output function because looking through tuples got annoying very fast. However they are a simpler function and as such it makes more sense to return them anywhere the output of the generator is needed

File: group/assignment6/tests.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+class TestMoveCars(unittest.TestCase):
+
+    def test_0(self):
+        self.assertEqual(return_output([1], [0]), ["Unsolvable"])
+        self.assertEqual(return_output([1, 2, 0], [1, 3, 0]), ["Unsolvable"])
+        self.assertEqual(return_output([1], [1]), [])
+        self.assertEqual(return_output([1, 2, 0], [1, 2, 0]), [])
+
+    def test_1(self):
+        self.assertEqual(return_output([1, 0], [0, 1]), ['Move car number 1 from 0 to 1'])
```
Comment (crossopt) [inline group/assignment6/tests.py:13]: Yeah, this is what I'm talking about. Why not check the result of the generator instead of weird strings?

File: group/assignment6/tests.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import unittest
+from move_cars import *
+
+class TestMoveCars(unittest.TestCase):
+
+    def test_0(self):
```
Comment (crossopt) [inline group/assignment6/tests.py:6]: Ooh, neat corner cases!

The names and split of functions are confusing though.
If test_0 is supposed to be test_incorrect, why are the solvable cases there?
Maybe move them to test_1, add a few more wrong input cases (empty arrays, different array sizes, several of the same cars)

Or just rename the test functions. I might have misunderstood you completely:)

File: group/assignment6/move_cars.py (hunk 1 above)
Comment (SilverSolver) [reply group/assignment6/move_cars.py:7]: I merged "car_positions" with "generate_moves" just because "car_positions" is called only once from "generate_moves".  So this merging makes the code shorter. I made validation the part of "generate_moves" for the same reason.
But I do not insist, because, yes, more functions can be more readable and testable, from the other side.
Let's see what others will say about separating code into logical blocks. I think that different approaches can exist there.

File: group/assignment6/move_cars.py (hunk 3 above)
Comment (SilverSolver) [reply group/assignment6/move_cars.py:43]: Yes, I think that checking Google code style is a good idea. I will do it and write a conclusion about suitable treatment of a wrong input etc in comments.
I also think that we should discuss this moment with other group members, as a part of program design.

File: group/assignment6/move_cars.py (hunk 4 above)
Comment (SilverSolver) [reply group/assignment6/move_cars.py:47]: Is is just a draft function which was written quickly and dirty and makes testing possible in some way. I agree that it should be redone later.

File: group/assignment6/tests.py (hunk 6 above)
Comment (SilverSolver) [reply group/assignment6/tests.py:6]: You are right, I'll reorganize test functions later.

File: group/assignment6/move_cars.py (hunk 1 above)
Comment (Sridha2412) [reply group/assignment6/move_cars.py:7]: having a different validation function could be a good idea. It will make it easy to read and test 👍

File: group/assignment6/move_cars.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                continue
+
+            if car:  # this is a cycle fix
+                yield car, place, pos[0]
+                pos[0], pos[car] = pos[car], pos[0]
+
+            car = final[pos[0]]
+            while car:
+                yield car, pos[car], pos[0]
+                pos[0], pos[car] = pos[car], pos[0]
```
Comment (Sridha2412) [inline group/assignment6/move_cars.py:34]: we cant swap two cars, it will have to be three moves

File: group/assignment6/move_cars.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                car = final[pos[0]]
+
```
Comment (Sridha2412) [inline group/assignment6/move_cars.py:36]: it doesn't return the correct final position

File: group/assignment6/tests.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        self.assertEqual(return_output([1], [0]), ["Unsolvable"])
+        self.assertEqual(return_output([1, 2, 0], [1, 3, 0]), ["Unsolvable"])
+        self.assertEqual(return_output([1], [1]), [])
+        self.assertEqual(return_output([1, 2, 0], [1, 2, 0]), [])
+
+    def test_1(self):
+        self.assertEqual(return_output([1, 0], [0, 1]), ['Move car number 1 from 0 to 1'])
+        self.assertEqual(return_output([1, 2, 0], [1, 0, 2]), ['Move car number 2 from 1 to 2'])
+
```
Comment (Sridha2412) [inline group/assignment6/tests.py:15]: Good tests! Might want to rename the function names thought 😅

File: group/assignment6/move_cars.py (hunk 7 above)
Comment (SilverSolver) [reply group/assignment6/move_cars.py:34]: pos[0] is not exactly the car, it is the empty space. It was discussed at the comments for crossopt's pull request for the same problem.

File: group/assignment6/move_cars.py (hunk 8 above)
Comment (SilverSolver) [reply group/assignment6/move_cars.py:36]: Can you give an example of the input or write the test which will show it?

**Comment by SilverSolver** [comment]
I'll better work in another branch from now on
